### PR TITLE
442 test bug in is equal

### LINKED ISSUE #442: Test Bug in is_equal
## Subject of the issue

We have been using `is_equal_tf` and `is_equal_np` for comparing numerics in tests.

However, the test might pass on CPU only machine if one value is nan and another value is zero.

But when running on GPU, the test failed: (this might be related to #187)

image

Change the functions into the following will fail the test.

```python
def is_equal_np(x: (np.ndarray, List), y: (np.ndarray, List)) -> bool:
    """return true if two numpy arrays are nearly equal"""
    x = np.asarray(x, dtype=np.float32)
    y = np.asarray(y, dtype=np.float32)
    return np.all(np.isclose(x, y))


def is_equal_tf(
        x: (tf.Tensor, np.ndarray, List), y: (tf.Tensor, np.ndarray, List)
) -> bool:
    """return true if two tf tensors are nearly equal"""
    x = tf.cast(x, dtype=tf.float32).numpy()
    y = tf.cast(y, dtype=tf.float32).numpy()
    return np.all(np.isclose(x, y))
```

## Your environment

- DeepReg version (commit hash or tag)

  8b8d75fdaaf89be2dfefc1d5c3c37e3ef26fd7d1

- OS (e.g. Ubuntu 20.04, MacOS 10.15, etc.)

  Ubuntu 18 Precision Tower

- Python Version (3.7, 3.8, etc.)

  3.7

- TensorFlow

  - TensorFlow Version 2.3
  - CUDA Version 10.1
  - cuDNN Version 7.6

### PR DESCRIPTION
**Pull request by mathpluscode**
# Description

Fixes #442

## Type of change

What types of changes does your code introduce to DeepReg? _Put an `x` in the boxes that
apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [ ] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      and formatted all changed files.
- [ ] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mathpluscode** [comment]
Tests on my XPS13 Ubuntu 20.04 is working but using GPU on PT failed with strange error:

```
test/unit/test_loss_image.py ....2020-10-07 00:30:22.060582: W tensorflow/core/framework/op_kernel.cc:1767] OP_REQUIRES failed at conv_ops_3d.cc:491 : Not found: No algorithm worked!                             
F                                                                                                                                                                                                                  
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>                                                                                                                                
                                                                                                                                                                                                                   
    def test_local_normalized_cross_correlation():                                                                                                                                                                 
        """                                                                                                                                                                                                        
        Testing computed local normalized cross correlation function between images using image.local_normalized_cross_correlation by comparing to precomputed.                                                    
        """                                                                                                                                                                                                        
        tensor_true = np.ones((2, 1, 2, 3, 2))                                                                                                                                                                     
        tensor_pred = 0.5 * np.ones((2, 1, 2, 3, 2))                                                                                                                                                               
        expect = [1, 1]                                                                                                                                                                                            
        get = image.local_normalized_cross_correlation(                                                                                                                                                            
            tensor_true,                                                                                                                                                                                           
>           tensor_pred,                                                                                                                                                                                           
        )                                                                                                                                                                                                          
                                                                                                                                                                                                                   
test/unit/test_loss_image.py:65:                                                                                                                                                                                   
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _                                                                                                                                
deepreg/model/loss/image.py:71: in local_normalized_cross_correlation                                                                                                                                              
    t_sum = tf.nn.conv3d(y_true, filters=filters, strides=strides, padding=padding)                                                                                                                                
../miniconda3/envs/deepreg/lib/python3.7/site-packages/tensorflow/python/util/dispatch.py:201: in wrapper                                                                                                          
    return target(*args, **kwargs)                                                                                                                                                                                 
../miniconda3/envs/deepreg/lib/python3.7/site-packages/tensorflow/python/ops/nn_ops.py:3033: in conv3d_v2                                                                                                          
    dilations, name)                                                                                                                                                                                               
../miniconda3/envs/deepreg/lib/python3.7/site-packages/tensorflow/python/ops/nn_ops.py:3006: in _conv3d_expanded_batch                                                                                             
    name=name)
../miniconda3/envs/deepreg/lib/python3.7/site-packages/tensorflow/python/ops/gen_nn_ops.py:1409: in conv3d
    _ops.raise_from_not_ok_status(e, name)
../miniconda3/envs/deepreg/lib/python3.7/site-packages/tensorflow/python/framework/ops.py:6843: in raise_from_not_ok_status
    six.raise_from(core._status_to_exception(e.code, message), None)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

value = None, from_value = None

>   ???
E   tensorflow.python.framework.errors_impl.NotFoundError: No algorithm worked! [Op:Conv3D]

<string>:3: NotFoundError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> <string>(3)raise_from()
```

I suspect it to be linked to Cuda problems

**Comment by mathpluscode** [comment]
By executing `export CUDA_VISIBLE_DEVICES=0` the tests are passed on PT machines.

Next step is to find a way to limit to at most 1 GPU for pytest.

**Comment by mathpluscode** [comment]
> By executing `export CUDA_VISIBLE_DEVICES=0` the tests are passed on PT machines.
> 
> Next step is to find a way to limit to at most 1 GPU for pytest.

It's nearly impossible to either
- set an environment variable for pytest without third party package;
- or limit the number of GPU devices once imported tf.

So instead, I changed the batch_size meaning in config so that batch_size in config means the size of mini-batch per GPU.

Tried pytest on PT with 3 GPUs, the test is passing ;)

File: deepreg/model/loss/label.py [hunk 1]
```diff
@@ -164,13 +162,11 @@ def dice_score(y_true: tf.Tensor, y_pred: tf.Tensor, binary: bool = False) -> tf
...
-    denominator = (
-        tf.reduce_sum(y_true, axis=[1, 2, 3])
-        + tf.reduce_sum(y_pred, axis=[1, 2, 3])
-        + EPS
+    numerator = tf.reduce_sum(y_true * y_pred, axis=[1, 2, 3]) * 2
+    denominator = tf.reduce_sum(y_true, axis=[1, 2, 3]) + tf.reduce_sum(
+        y_pred, axis=[1, 2, 3]
     )
-    return numerator / denominator
+    return tf.math.divide_no_nan(numerator, denominator)
```
Comment (YipengHu) [inline deepreg/model/loss/label.py:169]: isn't this dangerous for hiding potential bugs?

**Review by YipengHu — APPROVED** [review]

File: deepreg/model/loss/label.py (hunk 1 above)
Comment (mathpluscode) [reply deepreg/model/loss/label.py:169]: no, it's actually safer as 0/0 is not defined, so it can be 0 or anything.

**Pull request merged**